### ISSUE TITLE
Move Html helper to singleton.

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/yii2#10025]
Currently the static design of Html makes it hard to:
- Replace / override functionality.
- Use dependency injection.

By changing Html to be a normal class (and then register it with the DI container as a service) we can:
- Allow developers to selectively override functionality that is used by core components.
- Properly use dependency injection.

Since PHP allows calling static functions on objects, this change could be implemented without breaking BC as follows:

Change core code to use DI Container as Service Locator (Not ideal, but still better than hardcoded classname):
- global replace Html:: with \Yii::$container->get('Html')->

This way we can use the DI container to point to a different implementation of the Html Helper.

Of course ideally the helper class would not use static methods at all.

**Comment by samdark** [comment]
Replace/override isn't hard. There's LSB and a blank class that is easy to replace. What's the use of DI for HTML helper?

`\Yii::$container->get('Html')->` instead of  `Html::` looks totally ugly.

How would the syntax look in your ideal case?

**Comment by Alex-Code** [comment]
The helper classes are already simple to override.
https://github.com/yiisoft/yii2/blob/master/docs/guide/helper-overview.md#customizing-helper-classes-

**Comment by SamMousa** [comment]
What is LSB?

Well in an ideal perspective html helper would of course be injected via constructor or setter injection, so then syntax could be:

$this->html->

The issue is that currently the usage is hardcodes so now I have a folder in my project that contains an Html class in the yii\helpers namespace.
The issue is that I do not want to implement stuff in that namespace, ever ;-)

Also read http://kore-nordmann.de/blog/0103_static_considered_harmful.html for some background.

Basically the issue is "static dependencies are bad"; however I do not know what the best way is to solve this.
Some features that we should support:
1. App developer overrides certain functions in Html helper class that should be used by core.
2. Extension developer overrides certain functions in Html helper class that should be used by core and developer.
3. (Optional, but would be cool) Support some nice way of doing both.

**Comment by SamMousa** [comment]
I know Alex-Code, but that is quite an ugly construct in the age of namespacing and dependency injection.

Also it does not consider the use case where an extension extends the Html class as well; since now I cannot extend the extensions' class while also have the core components use that extended class.

**Comment by samdark** [comment]
LSB: https://secure.php.net/manual/en/language.oop5.late-static-bindings.php

**Comment by samdark** [comment]
I don't agree that static dependencies are worst solution in every possible case. They're bad if misused, of course but for the sake of syntax for something that's quite stable I think they're acceptable.

The features you've listed are supported. Extensions could deal with it using a trait.

**Comment by klimov-paul** [comment]
Duplicates #70
